File: python/v5.0.2023/M1/prime_logging.py
# prime v5.0.2023
# cycles through limits and writes to the filesystem

import math, time, os, cpuinfo

scope = [100, 1000, 10000, 100000, 1000000, 10000000, 25000000, 100000000, 1000000000]
reference = [25, 168, 1229, 9592, 78498, 664579, 1565927, 5761455, 50847534]
time_calc = [0, 0, 0, 0, 0, 0, 0, 0, 0]

def is_prime(number):
    global found
    flag_prime = 1
    for divider in range(3, int(math.sqrt(number)) + 1, 2):
        if number % divider == 0:
            flag_prime = 0
            break
    return flag_prime

def find_primes(largest):
    global primes
    global found
    for number in range(11, largest + 1, 2):
        if is_prime(number) > 0:
            found += 1
            primes.append(number)

def is_prime_fast(number):
    global found
    flag_prime = 1
    largest_divider = int(math.sqrt(number)) + 1
    for divider in primes:
        if number % divider == 0:
            flag_prime = 0
            break
        if divider > largest_divider:
            break
    return flag_prime

if __name__ == "__main__":
    for i in range(len(scope)):
        last = scope[i]
        found = 4              # we start from 11, know 2, 3, 5, 7
        primes = [3, 5, 7]     # exclude 2 since we only test odd numbers    
        print(f"Prime numbers to {last} in Python with algorithm v5.0.2023")
        start = time.perf_counter_ns()
        largest_divider = int(math.sqrt(last))
        if largest_divider % 2 == 0:
            largest_divider += 1
        print(f'First find prime dividers up to {largest_divider}.')
        find_primes(largest_divider)
        print(f'Found {found} primes, now use them als dividers.')
        for number in range(largest_divider + 2, last, 2):
            found += is_prime_fast(number)
        end = time.perf_counter_ns()
        print(f"This took: {(end - start)/1000} microseconds.")
        print(f"I found {found} prime numbers. Should be {reference[i]}.\n")
        filename = str(last) + ".txt"
        with open(filename, "w") as fp:
            fp.write('prime v5.0.2023 in python\n')
            fp.write(cpuinfo.get_cpu_info()['brand_raw'])
            fp.write(f'\nPrimes to {last} took {(end - start)/1000} microseconds.')
            fp.write(f'\nFound {found} primes. Should be {reference[i]}.\n')
        time_calc[i] = (end - start)/1000
    with open("summary.txt", "w")as fp:
        fp.write('Summary of prime v5.0.2023 in python\n')
        fp.write(cpuinfo.get_cpu_info()['brand_raw'])
        fp.write('\n     last       time in microseconds\n')
        for i in range(len(scope)):
            skip = " " * int(10.9 - math.log10(scope[i]))
            fp.write(skip)
            fp.write(f'  {scope[i]}   {time_calc[i]} \n')

File: c/v5.4.2024/M1/prime_logging.c
// prime to 2E32 version 5.4.2024 uint32 from 2023/12/23

#include <stdio.h>
#include <time.h>
#include <math.h>
#include <stdint.h>

int32_t primes[6550] = {3, 5, 7};    // to store the 6542 primes to sqr(2E32)
int32_t found = 4;                   // we already know 2, 3, 5, 7

int is_prime(int32_t number) {
    int prime = 1;
    for (int32_t divider = 3; divider < (int)(sqrt(number)) + 1; divider += 2)
    {
        if (number % divider == 0)
        {
            prime = 0;
            break;
        }
    }
    return prime;
}

void find_primes(int32_t limit) {
    for(int32_t number=11; number < limit + 1; number += 2) { 
        if( is_prime(number) == 1) {
            primes[found - 1] = number;
            found++;
        }
    }
    primes[found-1] = limit;
}

int is_prime_fast(uint32_t number)
{
    uint32_t largest_divider = (uint32_t)(sqrt(number));
    int flag_prime = 1;
    for(int i=0; i < found; i++)
    {
        if(number % primes[i] == 0) 
        {
            flag_prime = 0;
            break;
        }
        if(primes[i] >= largest_divider)
        {
            break;
        }
    }
    return flag_prime;
}


int main()
{
    const uint32_t scope[] = {100, 1000, 10000, 100000, 1000000, 10000000, 25000000, 100000000, 1000000000, 2147483647, 4294967295};
    const int reference[] = {25, 168, 1229, 9592, 78498, 664579, 1565927, 5761455, 50847534, 105097564, 203280221};
    for (int i = 0; i < 11; i++)
    {
        uint32_t last = scope[i];
        found = 4;
        clock_t start, end;
        double cpu_time_used;
        printf("Calculating prime numbers in C v5.4.2024 until %u \n", last);
        start = clock();
        int32_t largest_divider = (int32_t)(sqrt(last));
        if(largest_divider % 2 == 0)
        {
            largest_divider += 1;
        }
        find_primes(largest_divider);
        printf("Found %d primes until %d to use as dividers.\n", found, largest_divider);
        for(uint32_t number = largest_divider + 2; number < last; number += 2)
        {
            found += is_prime_fast(number);
        }
        end = clock();
        cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
        printf("Found %d prime numbers. It should be %d.\n", found, reference[i]);
        printf("This took %f seconds.\n\n",cpu_time_used);
        char filename[20];
        sprintf(filename, "%u.txt", scope[i]);
        FILE *logfile;
        logfile = fopen(filename, "w");
        fprintf(logfile, "Calculating prime numbers until %u in C v5.0.2023\n", last);
        fprintf(logfile, "Found %d prime numbers. It should be %d.\n", found, reference[i]);
        fprintf(logfile, "This took %f seconds.\n\n", cpu_time_used);
        fclose(logfile);
    }
}
